### DISCUSSION TITLE
Middleware problem with custom express server

### DISCUSSION THREAD

**Discussion by ost-ing** [vercel/next.js#33440] [Help]
### Run `next info` (available from version 12.0.8 and up)

 Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000
    Binaries:
      Node: 17.2.0
      npm: 8.1.4
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

17.2.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Other

### Describe the Bug

I've written some middleware using the new API to rewrite urls for fully translated routing within my app. I had it all working with a custom express server using version `12.0.5-canary.11`.

I then subsequently upgraded to `12.0.7` and now get this error:

```
Error: To use middleware you must provide a `hostname` and `port` to the Next.js Server
```

This check was introduced in [`12.0.5-canary.18` ](https://github.com/vercel/next.js/pull/31935/commits/a29a1fd624f17e92943c00b787fae2d1eaae52ce)

This problem also occurs with the most basic Middleware implementation:

```tsx
import type { NextRequest, NextFetchEvent } from 'next/server';
import { NextResponse } from 'next/server';

export function middleware(req: NextRequest, _ev: NextFetchEvent) {
  return NextResponse.next();
}
```

### Expected Behavior

Should be able to use middleware with custom server, with no error

### To Reproduce

Use a the express custom server with the basic middleware implementation

**Comment by NeoLSN** [comment]
I got the same issue.

**Comment by georgePadolsey** [comment]
For those confronting this issue that need a quick fix for the dev server and are following the [express custom server example](https://nextjs.org/docs/advanced-features/custom-server) or similar.

You can change your instantiation of app from:
```javascript
const app = next({ dev })
```

To something like
```javascript
const app = next({
    dev,
    hostname: "localhost",
    port: 3000 // or whichever port you're using
});
```

Hope this helps :)

**Comment by Ali-Hussein-dev** [comment]
@georgePadolsey When creating a custom server, it cannot be deployed to vercel [docs](https://nextjs.org/docs/advanced-features/custom-server), and I don't think the Vercel team wants to push users to use another server.

**Reply by georgePadolsey** [reply]
Very true, apologies -- to be clear I was just suggesting this as a quick work around for those that are already currently running a custom server and were seeking a solution as the current main docs don't mention it yet[^1] :)

[^1]: nor would I expect it to as middleware is currently [beta](https://nextjs.org/docs/messages/beta-middleware)
